### PR TITLE
Adding Interface and Gradient tests to `TrotterProduct`

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
This PR adds additional tests to the `TrotterProduct` class to ensure it integrates well with
all interfaces and is fully differentiable.

**Description of the Change:**
- Add tests for interface execution
- Add tests for interface gradient computation

### CODE REVIEW

File: tests/templates/test_subroutines/test_trotter.py [hunk 1]
```diff
@@ -469,3 +512,453 @@ def circ():
...
+                lambda x, y: x @ y,
+                [qml.matrix(op, wire_order=range(2)) for op in expected_product_sequence],
+            )
+            @ initial_state
+        )
+
+        state = circ(time, coeffs)
+        assert qnp.allclose(expected_state, state)
+
+    #   Gradient tests:
```
Comment (Jaybsoni) [inline tests/templates/test_subroutines/test_trotter.py:690]: These tests are currently failing, I have added the @xfail marker for now but they will be fixed before merging

**Comment by Jaybsoni** [comment]
[sc-44302]

File: tests/templates/test_subroutines/test_trotter.py (hunk 1 above)
Comment (soranjh) [reply tests/templates/test_subroutines/test_trotter.py:689]: Not sure what was the motivation to  use `_generate_grad_decomp` but I suggest performing gradient tests with one of the following methods:

1. Construct a manual expansion of `TrotterProduct` in a circuit and perform gradient calculations with a selected interface. The obtained gradients will be your reference and can be used for all tests, even those with different interfaces. For example, use the following reference circuit for applying `qml.TrotterProduct(h, time, n=1, order=1)`. 

```
@qml.qnode(dev)
def reference_circuit(time, coeffs):
    for coeff, op in zip(coeffs[::-1], terms[::-1]):
        qml.exp(op, coeff * (time / n) * 1j)
    return qml.expval(qml.Hadamard(0))

reference_gradients = qml.grad(circ)(time, coeffs)
```

2. Compute the gradients numerically with finite difference by slightly perturbing the Hamiltonian coefficients, computing the difference and dividing it by the perturbation.  This will be your reference gradient.

Comment (Jaybsoni) [reply tests/templates/test_subroutines/test_trotter.py:689]: I will do the former for now, but I think we should also ensure that the gradients we compute are actually correct. I don't think that simplify if the reference gradient and our gradient match then we have actually verified the gradients to be correct.

File: tests/templates/test_subroutines/test_trotter.py [hunk 2]
```diff
@@ -162,6 +163,47 @@
...
+        decomp.extend(qml.exp(op, coeff * (time / n) * 1j) for coeff, op in zip(coeffs, ops))
+
+    coeffs_ops = zip(coeffs, ops)
+
+    if get_interface(coeffs) == "torch":
+        import torch
+
+        coeffs_ops_reversed = zip(torch.flip(coeffs, dims=(0,)), ops[-1::-1])
+    else:
+        coeffs_ops_reversed = zip(coeffs[-1::-1], ops[-1::-1])
```
Comment (soranjh) [inline tests/templates/test_subroutines/test_trotter.py:180]: `coeffs[::-1]` does not do the same?

File: tests/templates/test_subroutines/test_trotter.py [hunk 3]
```diff
@@ -468,3 +511,330 @@ def circ():
...
+    @pytest.mark.jax
+    @pytest.mark.parametrize("time", (0.5, 1, 2))
+    def test_jaxjit_execute(self, time):
+        """Test that the gate executes correctly in the jax interface with jit."""
+        import jax
+        from jax import numpy as jnp
+
+        time = jnp.array(time)
+        c1 = jnp.array(1.23)
+        c2 = jnp.array(-0.45)
```
Comment (soranjh) [inline tests/templates/test_subroutines/test_trotter.py:558]: Why we cannot use `qml.dot(coeffs, terms)` here?

File: tests/templates/test_subroutines/test_trotter.py [hunk 4]
```diff
@@ -468,3 +511,330 @@ def circ():
...
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @jax.jit
+        @qml.qnode(dev, interface="jax")
+        def circ(time, c1, c2):
+            h = qml.sum(
+                qml.s_prod(c1, terms[0]),
+                qml.s_prod(c2, terms[1]),
+            )
```
Comment (soranjh) [inline tests/templates/test_subroutines/test_trotter.py:569]: Why we cannot use `qml.dot(coeffs, terms)` here?

File: tests/templates/test_subroutines/test_trotter.py [hunk 5]
```diff
@@ -468,3 +511,330 @@ def circ():
...
+
+            return qml.expval(qml.Hadamard(0))
+
+        with tf.GradientTape() as tape:
+            result = circ(time, coeffs)
+
+        measured_time_grad, measured_coeff_grad = tape.gradient(result, (time, coeffs))
+
+        with tf.GradientTape() as tape:
+            result = reference_circ(time, coeffs)
```
Comment (soranjh) [inline tests/templates/test_subroutines/test_trotter.py:800]: Is this the standard way to compute the gradients?

File: tests/templates/test_subroutines/test_trotter.py [hunk 6]
```diff
@@ -162,6 +163,47 @@
...
+
+        coeffs_ops_reversed = zip(torch.flip(coeffs, dims=(0,)), ops[-1::-1])
+    else:
+        coeffs_ops_reversed = zip(coeffs[-1::-1], ops[-1::-1])
+
+    if order == 2:
+        decomp.extend(qml.exp(op, coeff * (time / n) * 1j / 2) for coeff, op in coeffs_ops)
+        decomp.extend(qml.exp(op, coeff * (time / n) * 1j / 2) for coeff, op in coeffs_ops_reversed)
+
+    if order == 4:
```
Comment (soranjh) [inline tests/templates/test_subroutines/test_trotter.py:186]: Do we use this branch in any tests? (fine to keep it even if we do not use it, but we need to make sure it is accurate)

**Review by soranjh — APPROVED** [review]
Thanks @Jaybsoni! Left some questions for clarification and some minor comments.

File: tests/templates/test_subroutines/test_trotter.py (hunk 5 above)
Comment (Jaybsoni) [reply tests/templates/test_subroutines/test_trotter.py:800]: For TensorFlow yes

File: tests/templates/test_subroutines/test_trotter.py (hunk 2 above)
Comment (Jaybsoni) [reply tests/templates/test_subroutines/test_trotter.py:180]: Im not sure, I have always used `[-1::-1]`.

File: tests/templates/test_subroutines/test_trotter.py [hunk 7]
```diff
@@ -162,6 +163,47 @@
...
+    decomp = []
+    if order == 1:
+        decomp.extend(qml.exp(op, coeff * (time / n) * 1j) for coeff, op in zip(coeffs, ops))
+
+    coeffs_ops = zip(coeffs, ops)
+
+    if get_interface(coeffs) == "torch":
+        import torch
+
+        coeffs_ops_reversed = zip(torch.flip(coeffs, dims=(0,)), ops[-1::-1])
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply tests/templates/test_subroutines/test_trotter.py:178]: I think we would need to add a function to math to handle this, autoray didn't work when I tried it for both.

File: tests/templates/test_subroutines/test_trotter.py [hunk 8]
```diff
@@ -468,3 +511,330 @@ def circ():
         )
         state = circ()
         assert qnp.allclose(expected_state, state)
+
+    @pytest.mark.jax
+    @pytest.mark.parametrize("time", (0.5, 1, 2))
+    def test_jax_execute(self, time):
+        """Test that the gate executes correctly in the jax interface."""
+        import jax
```
Comment (Jaybsoni) [inline tests/templates/test_subroutines/test_trotter.py:519]:
```suggestion
        import jax  # pylint: disable=unused-import
```

File: tests/templates/test_subroutines/test_trotter.py (hunk 3 above)
Comment (Jaybsoni) [reply tests/templates/test_subroutines/test_trotter.py:558]: `qml.dot` is not fully compatible with jax jit. This allows us to by pass that restriction and still test that jax works with `TrotterProduct` by construction the hamiltonian manually using sum and scalar product.

File: tests/templates/test_subroutines/test_trotter.py (hunk 4 above)
Comment (Jaybsoni) [reply tests/templates/test_subroutines/test_trotter.py:569]: `qml.dot` is not fully compatible with jax jit. This allows us to by pass that restriction and still test that jax works with `TrotterProduct` by construction the hamiltonian manually using sum and scalar product.

File: tests/templates/test_subroutines/test_trotter.py (hunk 2 above)
Comment (soranjh) [reply tests/templates/test_subroutines/test_trotter.py:180]: [non-blocking] might check and verify.

File: tests/templates/test_subroutines/test_trotter.py (hunk 6 above)
Comment (soranjh) [reply tests/templates/test_subroutines/test_trotter.py:186]: We have tested order 4 for the decomposition, so no need to have it for gradients.

File: tests/templates/test_subroutines/test_trotter.py [hunk 9]
```diff
@@ -468,3 +511,330 @@ def circ():
         )
         state = circ()
         assert qnp.allclose(expected_state, state)
+
+    @pytest.mark.jax
+    @pytest.mark.parametrize("time", (0.5, 1, 2))
+    def test_jax_execute(self, time):
+        """Test that the gate executes correctly in the jax interface."""
+        import jax  # pylint: disable=unused-import
```
Comment (soranjh) [inline tests/templates/test_subroutines/test_trotter.py:519]: Can we just remove `import jax` instead of disabling pylint? I don't think we need it for this test.

Comment (Jaybsoni) [inline tests/templates/test_subroutines/test_trotter.py:519]:
```suggestion
```

File: tests/templates/test_subroutines/test_trotter.py (hunk 6 above)
Comment (Jaybsoni) [reply tests/templates/test_subroutines/test_trotter.py:186]: Added 4 decomposition to tests 👍🏼

File: tests/templates/test_subroutines/test_trotter.py (hunk 2 above)
Comment (Jaybsoni) [reply tests/templates/test_subroutines/test_trotter.py]: replaced 👍🏼
